src/jltypes.c

```diff
@@ -3635,43 +3358,50 @@ void jl_init_types(void)
     jl_set_typeof(jl_nothing, jl_void_type);
     jl_void_type->instance = jl_nothing;
 
-    jl_uniontype_type = jl_new_datatype(jl_symbol("Union"),
-                                        jl_type_type, jl_emptysvec,
-                                        jl_svec(1, jl_symbol("types")),
-                                        jl_svec(1, jl_simplevector_type),
-                                        0, 0, 1);
-
-    jl_bottom_type = (jl_value_t*)jl_new_struct(jl_uniontype_type, jl_emptysvec);
-
-    jl_tvar_type = jl_new_datatype(jl_symbol("TypeVar"),
-                                   jl_any_type, jl_emptysvec,
-                                   jl_svec(4, jl_symbol("name"),
-                                           jl_symbol("lb"), jl_symbol("ub"),
-                                           jl_symbol("bound")),
-                                   jl_svec(4, jl_sym_type, jl_type_type,
-                                           jl_type_type, jl_any_type),
+    jl_datatype_t *type_type = (jl_datatype_t*)jl_type_type;
+    jl_bottomtype_type = jl_new_datatype(jl_symbol("BottomType"), type_type, jl_emptysvec,
+                                         jl_emptysvec, jl_emptysvec, 0, 0, 0);
+    jl_bottom_type = jl_new_struct(jl_bottomtype_type);
+    jl_bottomtype_type->instance = jl_bottom_type;
+
+    jl_uniontype_type = jl_new_datatype(jl_symbol("Union"), type_type, jl_emptysvec,
+                                        jl_svec(2, jl_symbol("a"), jl_symbol("b")),
+                                        jl_svec(2, jl_any_type, jl_any_type),
+                                        0, 0, 2);
+
+    jl_tvar_type = jl_new_datatype(jl_symbol("TypeVar"), jl_any_type, jl_emptysvec,
+                                   jl_svec(3, jl_symbol("name"),
+                                           jl_symbol("lb"), jl_symbol("ub")),
+                                   jl_svec(3, jl_sym_type,
+                                           jl_any_type, jl_any_type),
                                    0, 1, 3);
 
+    jl_unionall_type = jl_new_datatype(jl_symbol("UnionAll"), type_type, jl_emptysvec,
+                                       jl_svec(2, jl_symbol("var"), jl_symbol("body")),
+                                       jl_svec(2, jl_tvar_type, jl_any_type),
+                                       0, 0, 2);
+
     vararg_sym = jl_symbol("Vararg");
     jl_svec_t *tv;
     tv = jl_svec2(tvar("T"),tvar("N"));
-    jl_vararg_type = jl_new_abstracttype((jl_value_t*)vararg_sym, jl_any_type, tv);
+    jl_vararg_type = (jl_unionall_t*)jl_new_abstracttype((jl_value_t*)vararg_sym, jl_any_type, tv)->name->wrapper;
```

Comment:
Valid types can no longer have free type variables: a `UnionAll` type must be present to bind each var. So `Vararg{T,N}` now has two layers of `UnionAll` around it. In the case of `Vararg` this is a bit awkward, since `Vararg` itself is not really a valid type, as it doesn't refer to a set of first-class values.